seqr/views/utils/dataset_utils.py

```diff
@@ -295,26 +294,106 @@ def match_and_update_samples(
+
+        yield sample_id, parsed
+
+def _check_invalid_tissues(samples, sample_id_to_tissue_type, warnings):
+    invalid_tissues = {}
+    for sample in samples:
+        tissue_type = TISSUE_TYPE_MAP[sample_id_to_tissue_type[sample.sample_id]]
+        if not sample.tissue_type:
+            sample.tissue_type = tissue_type
+            sample.save()
+        elif sample.tissue_type != tissue_type:
+            invalid_tissues[sample] = tissue_type
+
+    if invalid_tissues:
+        mismatch = ', '.join([
+            f'{sample.sample_id} ({REVERSE_TISSUE_TYPE[expected_tissue]} to {REVERSE_TISSUE_TYPE[sample.tissue_type]})'
+            for sample, expected_tissue in invalid_tissues.items()])
+        message = f'Skipped data loading for the following {len(invalid_tissues)} samples due to mismatched tissue type: {mismatch}'
+        warnings.append(message)
+
+    return [sample for sample in samples if sample not in invalid_tissues]
+
+def load_rna_seq_outlier(file_path, user=None, mapping_file=None, ignore_extra_samples=False):
+    expected_columns = ['sampleID'] + list(RNA_OUTLIER_COLUMNS.keys())
+    return _load_rna_seq(
+        RnaSeqOutlier, file_path, user, mapping_file, ignore_extra_samples, _parse_outlier_row, expected_columns,
+    )
+
+def load_rna_seq_tpm(file_path, user=None, mapping_file=None, ignore_extra_samples=False):
+    sample_id_to_tissue_type = {}
+    return _load_rna_seq(
+        RnaSeqTpm, file_path, user, mapping_file, ignore_extra_samples, _parse_tpm_row, TPM_HEADER_COLS,
+        additional_data=sample_id_to_tissue_type, validate_samples=_check_invalid_tissues,
+    )
+
+def _load_rna_seq(model_cls, file_path, user, mapping_file, ignore_extra_samples, parse_row, expected_columns,
+                  additional_data=None, validate_samples=None):
+    sample_id_to_individual_id_mapping = {}
+    if mapping_file:
+        sample_id_to_individual_id_mapping = load_mapping_file_content(mapping_file)
 
-def load_rna_seq(model_cls, file_path, user, sample_id_to_individual_id_mapping, ignore_extra_samples, parse_row, validate_header):
     samples_by_id = defaultdict(dict)
-    with gzip.open(file_path, 'rt') as f:
-        header = _parse_tsv_row(next(f))
-        validate_header(header)
-
-        for line in tqdm(f, unit=' rows'):
-            row = dict(zip(header, _parse_tsv_row(line)))
-            for sample_id, row_dict in parse_row(row):
-                gene_id = row_dict['gene_id']
-                existing_data = samples_by_id[sample_id].get(gene_id)
-                if existing_data and existing_data != row_dict:
-                    raise ValueError(
-                        f'Error in {sample_id} data for {gene_id}: mismatched entries {existing_data} and {row_dict}')
-                samples_by_id[sample_id][gene_id] = row_dict
+    f = file_iter(file_path)
+    header = _parse_tsv_row(next(f))
+    missing_cols = [col for col in expected_columns if col not in header]
```

Comment:
style - I find `missing_cols = set(expected_columns) - set(header)` easier to read.